store/structured/mongo/config.go

```diff
@@ -1,108 +1,88 @@
 package mongo
 
 import (
+	"context"
+	"fmt"
 	"net/url"
-	"strconv"
 	"strings"
 	"time"
 
-	mgo "github.com/globalsign/mgo"
+	"github.com/kelseyhightower/envconfig"
+	"go.mongodb.org/mongo-driver/x/mongo/driver/connstring"
 
-	"github.com/tidepool-org/platform/config"
 	"github.com/tidepool-org/platform/errors"
-	"github.com/tidepool-org/platform/pointer"
-)
 
-//Config describe parameters need to make a connection to a Mongo database
-type Config struct {
-	Scheme           string        `json:"scheme"`
-	Addresses        []string      `json:"addresses"`
-	TLS              bool          `json:"tls"`
-	Database         string        `json:"database"`
-	CollectionPrefix string        `json:"collectionPrefix"`
-	Username         *string       `json:"-"`
-	Password         *string       `json:"-"`
-	Timeout          time.Duration `json:"timeout"`
-	OptParams        *string       `json:"optParams"`
-}
+	"go.uber.org/fx"
+)
 
-//NewConfig creates and returns an incomplete Config object
-func NewConfig() *Config {
-	return &Config{
+func NewConfig(lifecycle fx.Lifecycle) *Config {
```

Comment:
I saw that you're passing a nil Lifecycle in some cases. To avoid this, I suggest splitting this to `NewConfig() *Config` which creates the struct and `AppendLifecycleHooksToConfig(config *Config, lifecycle fx.Lifecycle)` which appends the lifecycle hooks. You can then just call `NewConfig()` from the code that doesn't use `fx`. `NewStore(p Params) (*Store, error)` should also be split into two separate functions.

Here's an example:
```go

var StoreModule = fx.Options(
	fx.Provide(NewConfig),
	fx.Invoke(AppendLifecycleHooksToConfig),
	fx.Provide(NewStore),
	fx.Invoke(AppendLifecycleHooksToStore),
)


func NewStore(config *Config) *Store {
	return &Store{
		config: config,
		ctx:    context.Background(),
	}
}

func AppendLifecycleHooksToStore(store *Store, lifecycle fx.Lifecycle) {
	lifecycle.Append(fx.Hook{
		OnStart: func(ctx context.Context) error {
			return store.Initialize(ctx)
		},
		OnStop: func(ctx context.Context) error {
			return store.Terminate(ctx)
		},
	})
}


func NewConfig() *Config {
	return &Config{}
}

func AppendLifecycleHooksToConfig(cfg *Config, lifecycle fx.Lifecycle) {
	lifecycle.Append(fx.Hook{
		OnStart: func(ctx context.Context) error {
			if err := cfg.Load(); err != nil {
				return err
			}

			return cfg.Validate()
		},
	})
}
```